### ISSUE TITLE
Unbound variables in try blocks

### ISSUE THREAD

**Issue by sjrothfuss** [microsoft/pyright#5008]
**Describe the bug**
Pylance flags potentially unbound variables inside of `try` blocks.

**To Reproduce**
Type a variable that may not be bound into a `try` block. 

**Expected behavior**
Pylance should not flag potentially unbound variables in `try` blocks. The whole point of a `try` block is that the code might fail. I know I can add `# type: ignore` to the line but if I edit that line in the future I want Pylance to flag any real errors I introduce.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/108840392/234317591-de95c717-af17-4765-a7ea-194f5ec20f25.png)

**VS Code extension or command-line**
Running Pylance v2023.4.30 in VS Code

**Comment by erictraut** [comment]
Please provide a minimal, self-contained code sample in text form. Diagnosing issues from partial screen shots is not really possible.

**Comment by sjrothfuss** [comment]
I assumed this happens any time a potentially unbound variable is encountered so more detail besides "in a try block" was not needed. My apologies.

In the code below, Pylance flags the two references to `y` inside the `try` block with `"y" is possibly unbound`. So far as I know, the code runs without error for all values of `x`.

```python
def mynewfunction(x):
    if x == 5:
        y = [1,2,3]
    try:
        y.pop()
        print(y)
    except:
        print(x)

mynewfunction(1)
```

**Comment by erictraut** [comment]
Pyright is working as designed here.

It depends on what you mean by "runs with error". If you mean that the presence of an unbound variable is intentional and is caught by an exception handler, then yes, the code will run "without error", but this is a highly unusual coding pattern — one that I would discourage. You're using exception handling for basic code flow, not for the handling of exceptional conditions.

Here's a variant of your code that uses regular control flow statements:
```python
def mynewfunction(x):
    y = [1, 2, 3] if x == 5 else None

    if y:
        y.pop()
        print(y)
    else:
        print(x)
```

If you really want to use an exception handler to detect unbound variables at runtime, you can use `# type: ignore` to suppress these errors.

You can also completely disable the `reportUnboundVariable` for your entire project if you don't want pyright to check for this condition.
